### ISSUE TITLE
Concurrent Calls to SendMsg can hang

### ISSUE THREAD

**Issue by steve-gray** [grpc/grpc-go#2355]
### What version of gRPC are you using?
Latest

### What version of Go are you using (`go version`)?
1.11

### What operating system (Linux, Windows, …) and version?
Linux, Mac 

### What did you do?
1. Send multiple concurrent SendMsg calls from different goroutines to the same underlying GRPC stream.

### What did you expect to see?
Multiple calls to SendMsg should succeed, with no particular order enforced for concurrent calls.

### What did you see instead?
The calls ended up causing a hang, traced to the fact the write operations were not thread-safe, and multiple concurrent calls lead to inconsistent internal state for a send.

Was able to mitigate to a degree, by putting my own mutex around code that calls SendMsg - however when I then enable compression or have high latency, the SendMsg time now means that I spend more time waiting for compression than actually sending messages - despite this needing no particular blocking.

**Comment by dfawley** [comment]
Adding the same comment from PR #2356 for future reference:


Unfortunately, we do not want to enable concurrent `SendMsg` calls at this time.  There are a few reasons I can provide for this:

1. Because ordering often *does* matter for streams, it could be confusing for users if we allowed this.
1. Streams not attempting this type of operation become penalized by the cost of taking this mutex.
1. Similarly, in a heavily loaded system, attempting to send in parallel will not end up leading to any performance gains; the CPU and network are already fully saturated, and the cost of synchronization becomes a negative.
1. If we were to allow this, the issue of fairness arises pretty quickly thereafter.  `sync.Mutex`es have no concept of fairness, so it's possible that even if the user implements their own message reordering on the receiving end, one message could be postponed until many subsequent messages (hundreds/thousands/more) have been sent.  This could result in problems for users not expecting that possibility - e.g. if they are buffering a message until all prior messages are received.

If this is a performance bottleneck for your use case, I would recommend something like what @snowzach described in #1879.

**Comment by steve-gray** [comment]
@dfawley - seems like an odd position to take, given that the issue only exists because of mis-handing of concurrent operations against network buffers.

1. Order may matter, but it depends on use case - 'Unless you serialize calls, concurrent calls may arrive in an undefined order' - this is consistent with other language patterns and this allows scenarios such as order-required (serialise as today/no-change), order irrelevant (simply call in parallel), or semi-ordered (divide and serialise by some key). For example, if you publish in parallel to Kafka via Sarama, you get this behaviour.

2. The cost of the mutex is a single atomic memory check - and running the benchmarks before/after showed no appreciable cost to this - both with and without compression. What did you run to observe a meaningful impact so I can re-validate?

3. Are you able to articulate some more detail on this? It appears to defy my own understanding and experience, as well as the other changes in the issue you mentioned. With this change and calling from multiple routines, serialisation and compression can be run parallel - the locking is only to guard the HTTP/2 stream going out, and is sufficiently late in the process that instead of saturating a single core, I was able to saturate connections with both large and small messages, whereas previously enabling compression eventually consumes a core and caps throughput, and without it the ceiling is still there, but naturally higher, owing to the proto serialisation.

4. In terms of fairness, leaving this undefined at a specification level is simple enough - the scenario you're proposing is an artificially complex one - and its an application level consideration if they want to multiplex concurrent sets of data into a single stream - and they can still achieve that by not multiplexing writes on the producer side (i.e. nothing breaks)

The workaround in #1879 puts enough of the framework level overhead into the per-application logic, and renders it difficult to maintain a cross-language producer/consumer - as now you're bypassing core features of the framework. Today the framework hangs, deadlocking the callee goroutines if you perform concurrent writes with no application level way to recover or even differentiate this has occured.

**Comment by snowzach** [comment]
@suyashkumar I had the same issue. When I was sending large messages, the compress and encoding actually took a lot longer than the writing. Check out my comment in https://github.com/grpc/grpc-go/issues/1879 for what I did. I was able to saturate my network link after that.

**Comment by suyashkumar** [comment]
@snowzach will do, thanks!

**Comment by dfawley** [comment]
@steve-gray 

That `SendMsg` is not thread-safe is not an oversight.  The use case of unordered messages on a stream is uncommon.  Adding synchronization to support a rare use case is not worth the tradeoffs that it entails.  And it also doesn't help with cases like @suyashkumar's, where ordering is necessary, but the same limitations prevent maximum throughput.

If your messages are order-independent, you could also try creating multiple streams to carry the messages.  This will actually scale better long-term as well, in case you fully saturate your connection and find you need multiple connections to provide even more throughput.

For cases where ordering does matter, solutions like @snowzach's are the only way to help, currently.

Note that the writing of the data to the connection doesn't actually happen in the `SendMsg` goroutine (as of ~1yr ago).  The data is queued and a centralized goroutine picks it up and writes it.  The `SendMsg` goroutine only blocks until sufficient flow control is available for the queue before returning.  This means that, if your messages are large, you typically will be in the N+1th call to `SendMsg` while the data from the Nth call is being written.  If you can marshal/compress the data faster than you can write it, then that will saturate the network.  If it's slower, the marshal/compress would need to happen in multiple goroutines to saturate the network.

In general, we believe that resolving these types of performance problems is best left up to the application.  Adding extra parallelism/synchronization in grpc incurs a penalty to users that do not require it.  That said, we're sensitive to applications needing to implement complex logic to get the most out of grpc, so where new APIs or libraries can be provided to help with that, we may do so.

**Comment by dfawley** [comment]
I filed issue #2432 as a proposal for a way to address this limitation.  Please feel free to comment.
